storage/shard.go

```diff
@@ -762,21 +814,36 @@ func (s *dbShard) Write(
 		// as the commit log need to use the reference without the
 		// overhead of ownership tracking. This makes taking a ref here safe.
 		commitLogSeriesID = entry.series.ID()
+		commitLogSeriesTags = entry.series.Tags()
 		commitLogSeriesUniqueIndex = entry.index
-		entry.decrementReaderWriterCount()
 		if err != nil {
+			entry.decrementReaderWriterCount()
 			return err
 		}
+		needsIndex := indexWriteFn != nil && entry.needsIndexUpdate(timestamp)
```

Comment:
NB: this function (`writeAndIndex`) is pretty involved so I tried breaking it into smaller pieces, can see the changes it'd take [here](https://github.com/m3db/m3db/commit/668bd3e92b1b40a3634a469ac89d1a8429b6f1cb). 

But that suffers from poor in-lining. Even with `-l=4` it still won't inline any function that calls an interface so it'd affect performance. How much impact I don't know. Can test impact if people think it's worth going down the road.